Fragility Algo. Added As Analyzer

adam2392 opened the pull request:
Added the fragility analysis algo. as an analyzer. The demo notebook is in corresponding pull request in tvb-documentation within demos directory.

This allows one to analyze the corresponding output time series generated by the TVB either in a serial fashion, or parallel (with multiprocessing pool).

maedoc committed:
add noise to base monitor

adam2392 committed:
OBSERVATION NOISE

Adding observation noise. Can be used as a keyword argument into
any Projection subclass.

For example, user sets an obsnoise object from the
Additive, or Multiplicative class:
            mon_SEEG = monitors.iEEG.from_file(period=period,
                                           variables_of_interest=[1],
                                           obsnoise=obsnoise)

To Do:
1. Check with TVB member that this works as intended. Works rn
on my local machine.
2. Does not generate line noise. Needs to be generated at a specific
frequency bands.
3. Needs to check dt, shape terms within Projection sample function.
4. How should nsig be determined. Should it have size (1,), or (numchans,)
since we are adding noise to the SEEG channels?

adam2392 committed:
INITIAL COMMIT FOR FRAGILITY

Created fragility analysis using the framework of TVB datatypes.
Took away all unecessary code from my own library, and only
included the code necessary to run things.

Now, the code is being tested for correctness and stability.

adam2392 committed:
Merge branch 'trunk' of https://github.com/adam2392/tvb-library into fragilitydev

adam2392 committed:
TESTED FRAGILITY ALGORITHM SERIAL COMPUTATION

Using a notebook demo, tested and computed the fragility end to end
after a test TVB simulation of 10 seconds.

To Do:
1. Test parallel computation and where to save temporary data and merging.

adam2392 committed:
Added abbility for fragility to add line noise.

adam2392 committed:
Adding the temporary directory creation using the python module as suggested by Maedoc. Also added the TVB logger to each of the execute files (serial and multiprocessing.

adam2392 commented:
The corresponding notebook demo will be added to this directory when the tvb-documentation is merged. 

https://github.com/the-virtual-brain/tvb-documentation/tree/master/demos

Let me know if there is anything else I need to add.

maedoc commented on tvb/analyzers/fragility.py:
```diff
@@ -0,0 +1,172 @@
+
+    # whether or not to use multiprocessing
+    parallelize = False
+    numcores = psutil.cpu_count() / 2.
```
`psutil` is a soft dependency, please use `multiprocessing.cpu_count()`, which ships with Python

adam2392 replied:
Fixed.

maedoc commented on tvb/analyzers/fragility.py:
```diff
@@ -0,0 +1,172 @@
+        "Returns the resulting model."
+        n = self.adjmats.shape 
+        return n
+    def result_size(self):
```
single line between methods please

adam2392 replied:
Fixed.

maedoc commented on tvb/analyzers/fragility.py:
```diff
@@ -0,0 +1,172 @@
+            
+            # create temp dir to save files
+            tempdir = None
+            assert tempdir is str
```
there's a problem here: `tempdir` will be `None` not `str`

adam2392 replied:
Fixed assertion statement.

maedoc commented on tvb/analyzers/fragility.py:
```diff
@@ -0,0 +1,172 @@
+
+    # whether or not to use multiprocessing
+    parallelize = False
+    numcores = psutil.cpu_count() / 2.
```
as above, please use builtin `multiprocessing.cpu_count()`

adam2392 replied:
Fixed.

maedoc commented on tvb/analyzers/fragility.py:
```diff
@@ -0,0 +1,172 @@
+        """
+        To be implemented in every subclass.
+        """
+        return None
```
this can return just a simple dictionary liek `{"radius": self.radius}`

adam2392 replied:
Added return statements for both the linear model estimation and the perturbation model.

maedoc commented on tvb/analyzers/fragility_lib/execute/singlecore/singlemvar.py:
```diff
@@ -0,0 +1,95 @@
+import os
+sys.path.append('../../')
+from ...linearmodels.basemodel import BaseWindowModel
+from ...linearmodels.mvarmodel import MvarModel
```
modifying `sys.path` isn't appropriate; you're probably just missing some `__init__.py` files in the relevent folders.  Let me know if I should take a look

adam2392 replied:
Yes, you are right. I removed this line. I think the 

"from ...linearmodels.basemodel ..." 

should handle it correctly.

maedoc commented on tvb/analyzers/fragility_lib/execute/singlenode/parallelmvar.py:
```diff
@@ -0,0 +1,136 @@
+    # 2: call the function to create the mvar model
+    adjmat = shared_mvarmodel.mvaradjacencymatrix(eegwin)
+        
+    tempfilename = os.path.join(shared_tempdir, 'temp_'+str(win)+'.npz')
```
`'temp_{0}.npz'.format(win)`

adam2392 replied:
Fixed these string formatters.

maedoc commented on tvb/analyzers/fragility_lib/execute/singlenode/parallelmvar.py:
```diff
@@ -0,0 +1,136 @@
+
+class ParallelMvar(BaseWindowModel):
+    def __init__(self, winsizems=250, stepsizems=125, samplerate=1000,numcores=None):
+        BaseWindowModel.__init__(self, 
```
`super(ParallelMvar, self).__init__(...)` is idiomatic because it handles multiple inheritance correctly.

adam2392 replied:
Updated to be super(<classname>, self) for each file in the execute dir.

maedoc commented on tvb/analyzers/fragility_lib/execute/singlenode/parallelmvar.py:
```diff
@@ -0,0 +1,136 @@
+            mvarresults = p.map(mvarjobwin, range(0, self.numwins))
+        p.join()
+        # mvarresults = [1]
+        sys.stdout.write('Finished mvar model computation for ' +\
```
should this be a logging statement instead, e.g. 

`logger.debug('Finsih comp for %d windows', self.numwins)`

maedoc commented on tvb/simulator/monitors.py:
```diff
@@ -514,6 +521,21 @@ def sample(self, step, state):
+            # By: Adam Li - Needs to Be Checked
+            # add observation noise if available
+            '''
+            1. Should dt be the same as in the source level?
```
isn't this change done in the other PR? if so, handle my comments there on that branhc and then we can rebase this one on that, before merging this PR

adam2392 replied:
Yes, I think I made a mistake when I was modifying files on this branch. I reverted the file back to what is on the trunk of the master branch. 

I updated the other branch (obs-noise) with updates to this.

maedoc reviewed (changes requested):
looks good, but please try to clean up the style per my comments also consider running autopep8 on the files you've added, and then we'll merge.

adam2392 committed:
Revert the monitors file back to the original master... I think it got synced by accident when I was sending a pull request for the fragilitydev branch.

adam2392 committed:
UPDATED FOR AUTOPEP8 AND SOME FORMATTING BUG FIXES

Updated any subclasses to use the idiomatic syntax for super init.
Ran autopep8 on the entire library. Removed unecessary lines and package
imports for the models.
